### ISSUE TITLE
A weird auto-use of block formation

### ISSUE THREAD

**Issue by Day-Elven** [genbtc/AutoTrimps#55]
http://screenshotlink.ru/dd66f7e2c23baa4b705a7091c79adbc4.png
I have seen this bug many times on high zones, and it gets most destructive when you need to reach higher zone than usual (in my case it is bionic 395). In other words, this bug hinders progress on high zones (it seems that it activates when monsters kills trimps faster than trimps breed, i.e. when trimps stay alive for less than 30 seconds my case). While my block is much higher than rivals damage, script still uses block formation, which makes no sense. Block formation reduce incoming damage by 2 times (because opponent's block pierce subsides from 15% to 7,5%), but decreases damage by 8 times, i.e. it makes your progress at least 4 times slower. Apparently, Dominance formation would be the most efficient.

**Comment by genbtc** [comment]
AutoStance 1 or 2 or both ? I think 2 does that, and I think I have found the issue through unrelated means.

But next time you see this, pause, please export your game and pastebin a copy to here

**Comment by genbtc** [comment]
OK I put in a patch. See if it works, i couldnt test it.

**Comment by Day-Elven** [comment]
It was Autostance 1, I didn't try Autostance 2. I'm going to try it now.

**Comment by Day-Elven** [comment]
On autostance 2:
Now it does not use Block formation all the time.
But if your trimps have very small amount of HP (less than 5% maybe), [and if the next trimps group is not ready], it still often behaves in a strange way. There are two types of strange behaviour:
1. It switches to block.
2. It starts to switch between several stances in a chaotic way and very fast speed, and finally stops on block formation.

**Comment by genbtc** [comment]
find such behavior save and export your save and post it to pastebin

**Comment by Day-Elven** [comment]
I could not catch it one more time. Now it use only D, X and H formations on high-level zones.

**Comment by genbtc** [comment]
Another guy found it: https://www.reddit.com/r/Trimps/comments/5ksrs0/problem_with_autostance_on_corrupted_precision/

**Comment by CherryInHove** [comment]
Hi,

I mentioned on reddit that I had this problem - I've attached the game pasued where you can see it is using Scryer stance on a corrupted even though I have it set to never use Scryer on corrupted. This sometimes also uses Barrier formation on corrupted as this thread mentioned which is almost never the best setting to use.

[Trimps Exported Game.txt](https://github.com/genbtc/AutoTrimps/files/676973/Trimps.Exported.Game.txt)

[Trimps Stance.txt](https://github.com/genbtc/AutoTrimps/files/676974/Trimps.Stance.txt)

**Comment by genbtc** [comment]
These files are identical, please include an export the AT settings you used.

Analysis:
I've figured out whats going on. 
This is slightly different than the OP's (Day-Elvens') problem - but the same as the reddit thread. 

In the game you sent me, (you are on S stance) https://puu.sh/thjWI/41ae1a3384.png  (full disclosure I gave the enemy's his full HP back to test cause you already had hit him a bunch)

You're right, this should not be fought in S stance. The cell before it was a Magma Cell. Therefore it used S for that one, and when it reached this cell, Scryer code DID get disabled, it reverted to Autostance code (since they are actually 2 different code bases), but Autostance expected you be already in D,X or B stance. Therefore it did not switch. This is indeed a bug.
Fighting this same fight again when D is pre-selected, correctly switches it to B stance (Which is where it should be) https://puu.sh/thk55/f00253f809.png

TLDR: There needs to be a code pathway to allow Autostance to switch into B when its in S, and I can do this patch soon.

As for "This sometimes also uses Barrier formation on corrupted as this thread mentioned which is almost never the best setting to use."

Now try analyzing the numbers of the last screenshot (or compare with the first) - I honestly don't agree with that statement at all. B is the correct formation to use, "Corrupted Precision" is what triggers it, due to the chance of 400% Crit extra damage. Even before a crit, the enemy is already doing 1.07e75 dmg * (1+400%/100) = 5.35e75 dmg on crit. You can see you only have 1.68e75 health so you would be insta-killed, (greatly wasting the breed timer and just not worth dying for) but with Block you do have the sufficient block (barely adequate at 5.42e75). So you survive just fine.

Unless there is something else going on which very well is possible. 

NOTE: Type in "notation" into the trimps itself's settings and change to Scientific Notation to stop using stuff like QaV DcD AsD FkG nobody knows what that means lol

**Comment by genbtc** [comment]
I put some fixes in. Try again. Sorry for long delay.

**Comment by CherryInHove** [comment]
Perfect! Thanks very much for your help on this, working perfectly now.

**Comment by CherryInHove** [comment]
Although one thing with regards to using H Vs using B:

Now try analyzing the numbers of the last screenshot (or compare with the first) - I honestly don't agree with that statement at all. B is the correct formation to use, "Corrupted Precision" is what triggers it, due to the chance of 400% Crit extra damage. Even before a crit, the enemy is already doing 1.07e75 dmg * (1+400%/100) = 5.35e75 dmg on crit. You can see you only have 1.68e75 health so you would be insta-killed, (greatly wasting the breed timer and just not worth dying for) but with Block you do have the sufficient block (barely adequate at 5.42e75). So you survive just fine.

Unless there is something else going on which very well is possible.

You're correct, however as 20% of the damage will penetrate block, which means that if they crit with block on, I'm going to be insta killed anyway, so I think that generally I'd be more likely to survive a couple of (non crit) hits with H on than with B, but I could be entirely wrong about this.

This  may just be relevant to the specific period of the game I'm in though and not affect anyone else, so probably isn't worth time trying to fix.

**Comment by MikkelVilla** [comment]
I don't know if this is the correct place to report it but I've notice some issue with auto-stance 1 and 2.

When I run void maps, specifically the 'henious' type, auto stance will swap from domination when the health gets low and into barrier formation. It will never use no formation of heap. It does this despite the fact that my block is 0 (my highest zone reached is 135 so block is irrelevant).

Is there any reason for this and is there some way I can stop it?
